pkg/action/failover.go

```diff
@@ -0,0 +1,384 @@
+package action
+
+import (
+	"fmt"
+	"slices"
+	"time"
+
+	storkv1 "github.com/libopenstorage/stork/pkg/apis/stork/v1alpha1"
+	"github.com/libopenstorage/stork/pkg/log"
+	"github.com/libopenstorage/stork/pkg/resourceutils"
+	"github.com/libopenstorage/stork/pkg/utils"
+	coreops "github.com/portworx/sched-ops/k8s/core"
+	storkops "github.com/portworx/sched-ops/k8s/stork"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+)
+
+// validationsBeforeFailover This method is called as part of the Initial stage of Failover
+func (ac *ActionController) validationsBeforeFailover(action *storkv1.Action) {
+	// TODO: source cluster-domain deactivation in sync-dr case
+	if action.Status.Status == storkv1.ActionStatusSuccessful {
+		// Move to ScaleUpDestination stage directly in case skipDeactivateSource is true
+		if *action.Spec.ActionParameter.FailoverParameter.SkipDeactivateSource {
+			logEvents := ac.printFunc(action, string(storkv1.ActionStatusSuccessful))
+			logEvents("Skipping the deactivation of source cluster and moving to ScaleUpDestination stage since --skip-deactivate-source is provided", "out")
+			action.Status.Stage = storkv1.ActionStageScaleUpDestination
+		} else {
+			action.Status.Stage = storkv1.ActionStageScaleDownSource
+		}
+		action.Status.Status = storkv1.ActionStatusInitial
+		action.Status.FinishTimestamp = metav1.Now()
+		ac.updateAction(action)
+		return
+	} else if action.Status.Status == storkv1.ActionStatusFailed {
+		action.Status.Stage = storkv1.ActionStageFinal
+		action.Status.FinishTimestamp = metav1.Now()
+		ac.updateAction(action)
+		return
+	}
+
+	// first get the migrationSchedule's static copy in the destination cluster
+	migrationSchedule, err := storkops.Instance().GetMigrationSchedule(action.Spec.ActionParameter.FailoverParameter.MigrationScheduleReference, action.Namespace)
+	if err != nil {
+		msg := fmt.Sprintf("error getting migrationschedule %s/%s for failover", action.Namespace, action.Spec.ActionParameter.FailoverParameter.MigrationScheduleReference)
+		logEvents := ac.printFunc(action, string(storkv1.ActionStatusFailed))
+		logEvents(msg, "err")
+		action.Status.Status = storkv1.ActionStatusFailed
+		action.Status.Reason = msg
+		ac.updateAction(action)
+		return
+	}
+	// get sourceConfig from clusterPair in the destination cluster
+	sourceConfig, err := getClusterPairSchedulerConfig(migrationSchedule.Spec.Template.Spec.ClusterPair, migrationSchedule.Namespace)
```

Comment:
Don't we have to consider "skip-source-actions" flag?

Reply:
I have not yet updated cli and controller according to the new discussion around skip flag. 
I will take that up in a separate PR.
For now, I have enforced clusterpair for everyone. As part of a susbsequent PR I can modify it such that we skip it if skip-source-operations flag is true

Reply:
Just to clarify what all is left even after this PR.
1. Cluster domain deactivate/activate and operations specific to sync-dr usecase.
2. Fixing the updateReplicas method used to scale up apps so that both reactivate and activate work as expected.
3. Design change discussed earlier this week around --skip-source-operations flag (this will require changes in cli as well)

Reply:
Have the changes for point 2 ready and will raise that PR once this PR is merged.
Will work on 1 and 3 early next week and raise the PR asap